profiler/test/Datadog.Profiler.Native.Tests/LinkedListTest.cpp

```diff
@@ -0,0 +1,320 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+// Our linked list implementation allows the allocator to return nullptr. This is ok in our case
+// because, we would use a ringbuffer-based allocator which might return nullptr.
+// To avoid UBSAN from failing our test, we deactivate the attribute nonnull-returns found
+// in memory_resource file (linux only)
+#pragma clang attribute push(__attribute__((no_sanitize("returns-nonnull-attribute"))), apply_to = function)
+#include "LinkedList.hpp"
+#pragma clang attribute pop
+
+#include "gtest/gtest.h"
+
+TEST(LinkedListTest, PushBack)
+{
+    LinkedList<int> x;
+
+    ASSERT_EQ(x.Size(), 0);
+
+    ASSERT_TRUE(x.Append(42));
+
+    ASSERT_EQ(x.Size(), 1);
+
+    ASSERT_EQ(*(x.begin()), 42);
+}
+
+TEST(LinkedListTest, ForEach)
+{
+    LinkedList<std::string> x;
+
+    ASSERT_TRUE(x.Append("1"));
+    ASSERT_TRUE(x.Append("2"));
+    ASSERT_TRUE(x.Append("3"));
+
+    int i = 1;
+
+    for (auto const& s : x)
+    {
+        ASSERT_EQ(std::to_string(i++), s);
+    }
+
+    ASSERT_EQ(x.Size(), 3);
+}
+
+struct Person
+{
+    std::string Name;
+    std::uint8_t Age;
+};
+
+TEST(LinkedListTest, Move)
+{
+    LinkedList<Person> l;
+
+    l.Append({.Name = "Georges", .Age = 42});
+    l.Append({.Name = "Ralph", .Age = 101});
+    l.Append({.Name = "jordy", .Age = 21});
+    l.Append({.Name = "Bob", .Age = 1});
+
+    ASSERT_EQ(l.Size(), 4);
+
+    LinkedList<Person> other = std::move(l);
+
+    ASSERT_EQ(0, l.Size());
+
+    ASSERT_EQ(other.Size(), 4);
+
+    auto it = other.begin();
+
+    ASSERT_EQ((*it).Name, "Georges");
+    ASSERT_EQ((*it).Age, 42);
+
+    ++it;
+
+    ASSERT_EQ((*it).Name, "Ralph");
+    ASSERT_EQ((*it).Age, 101);
+
+    ++it;
+
+    ASSERT_EQ((*it).Name, "jordy");
+    ASSERT_EQ((*it).Age, 21);
+
+    ++it;
+
+    ASSERT_EQ((*it).Name, "Bob");
+    ASSERT_EQ((*it).Age, 1);
+}
+
+TEST(LinkedListTest, Assign)
+{
+    LinkedList<Person> l;
+
+    l.Append({.Name = "Georges", .Age = 42});
+    l.Append({.Name = "Ralph", .Age = 101});
+    l.Append({.Name = "jordy", .Age = 21});
+    l.Append({.Name = "Bob", .Age = 1});
+
+    ASSERT_EQ(l.Size(), 4);
+
+    LinkedList<Person> other;
+
+    other = std::move(l);
+
+    ASSERT_EQ(l.Size(), 0);
+
+    auto begin_it = l.begin();
+    auto end_it = l.end();
+    ASSERT_EQ(begin_it, end_it);
+
+    // check other now
+    ASSERT_EQ(other.Size(), 4);
+
+    auto it = other.begin();
+
+    ASSERT_EQ((*it).Name, "Georges");
+    ASSERT_EQ((*it).Age, 42);
+
+    ++it;
```

Comment:
I know that ++iterator is using iterator++ at the implementation level but maybe check that iterator++ is also working - could be in this same test method

Reply:
I use the prefix increment in one test and the postfix in another one